### ISSUE TITLE
Prepared statements are not actually executed.

### ISSUE THREAD

**Issue by bitifet** [brianc/node-postgres#3295]
### Behaviour observed

According to [documentation](https://node-postgres.com/features/queries#prepared-statements), if the name parameter is supplied to the query config object it should:

  * Prepare the statement given in the *text* property.
  * Execute it.
  * Use it next time *.query()* method is called with the same name.

But, from my observation, what happen is:

  * ✅ It is effectively prepared.
  * ✅ If you call *.query()* method again with a different string in *text* property it throws an error.
  * ❌ But what is executed in the server is not the prepared statement but the originally provided text for preparation.
  * ❌ Same in subsequent calls, even if the *text* property is omitted.

### Tests done

Implementing an automated test to prove it is hard because we cannot send sentences through a client while a command is executed through it and from a different client we cannot see what is happening in other sessions.

But I wrote down following script that allow to verify what I am saying looking at [pg_activity](https://github.com/dalibo/pg_activity/):

```
const {Client} = require("pg"); 
const client = new Client();
const sql = "select pg_sleep(5)"; // Just give time to see it in pg_activity
const sleep = (ms)=>new Promise(resolve=>setTimeout(resolve, ms));
const banner = msg=>client.query(`select '${msg}' from pg_sleep(5)`);

client.connect().then(async function prepare_playground() {
    
    console.log("Look at pg_activity...");
    
    void await banner("Executing unnamed...");
    void await client.query({text: sql});
    
    void await banner("Executing named... (gets prepared)");
    void await client.query({name: "named", text: sql});
    
    void await banner("Execute expected but only same sql send to the server");
    void await client.query({name: "named"});
    
    void await banner("Actual execution");
    void await client.query({text: "execute named"});
    
    client.end();
    console.log("Done!");
    
});
```

You can repeat the test with the following steps:

  * Paste former code to some js file (let's say *prepare_test.js*).
  * Install node_postgres: `npm init -y && npm install pg`
  * Open two terminals and set needed environment variables to connect a database.
  * Open *pg_activity* in one of the terminals **(ensure it is wide enough to show the *Query* column)**.
  * Execute the script in the other: `node prepare_test.js`.
  * Look at the *Query* column of pg_activity

Expected output (you will see following lines in the *Query* column sequentially one after the other):

```
select 'Executing unnamed...' from pg_sleep(5)
select pg_sleep(5)
select 'Executing named... (gets prepared)' from pg_sleep(5)
select pg_sleep(5)
select 'Execute expected but only same sql send to the server' from pg_sleep(5)
select pg_sleep(5)
select 'Actual execution' from pg_sleep(5)
execute named
```

Notice that only the last command actually executes the prepared statement.

This proves:

  1. It's actually prepared.
  2. That preparation is not (automatically) used as expected.

**Comment by brianc** [comment]
Been reading this over a few times & unfortunately I am having trouble following what you are saying.

**Comment by brianc** [comment]
> * But what is executed in the server is not the prepared statement but the originally provided text for preparation.

What does this mean?

> * Same in subsequent calls, even if the _text_ property is omitted.

What does this mean?

**Comment by bitifet** [comment]
You are absolutely right!

I apologize for wasting your time...

I already had checked that that `connection.parse(...)` statement were executed only once. But I also observed that it throws if I try to pass an already used *name* with different string in *text* so I guessed it is cached to compare and it looked reasonable that it could be taken and sent again by mistake.

I didn't know anything about Postgres communication protocol and trusted excessively in what *pg_activity* showed me.

Now I put a man in the middle and checked that there is a different command to trigger prepared statements than for regular SQL:

```
Data from client: Q3select 'Executing unnamed...' from pg_sleep(5)
Data from client: Qselect pg_sleep(5)
Data from client: QAselect 'Executing named... (gets prepared)' from pg_sleep(5)
Data from client: Pnamedselect pg_sleep(5)
Data from client: Bnamed
Data from client: DPE   S
Data from client: QTselect 'Execute expected but only same sql send to the server' from pg_sleep(5)
Data from client: BnamedDP
Data from client: E
Data from client: S
Data from client: Q/select 'Actual execution' from pg_sleep(5)
Data from client: Qexecute named
Data from client: X
```
...but typing the "execute" command in psql it obviously send that command to the server instead of interpreting it and using a different communication pattern. And hence I saw the difference and misinterpreted it.

Thank you for your fast response and, again: excuse me for the inconvenience.
